### PR TITLE
Add worker multiple receive action support

### LINKED ISSUE #18163: [Spec Deviation] "Multiple receive action" is not supported
**Description:**
See the "Multiple receive action" section of the spec.

Can't get the following to work
```ballerina
import ballerina/io;

public function main() {
    worker w1 {
        100 -> w2;
    }

    worker w2 {
        map<int> m = <- { w1 };
        io:println("m: ", m);    
    }

    wait w2;
}
```

**Affected Versions:**
```cmd
$ ballerina -v
Ballerina 1.0.0-alpha3
Language specification 2019R2
Ballerina tool 1.0.0
```

### PR DESCRIPTION
**Pull request by sanjana**
Fixes #18163

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -1826,6 +1832,258 @@ public void visit(BLangWorkerReceive workerReceiveExpr) {
...
+        }
+        // return boolean depending on whether a semantic error
+        return noSemanticErrors;
+    }
+
+    private boolean checkTypeForWorkerMultipleMap(WorkerMultipleReceiveNode.WorkerReceiveFieldNode receiveField,
+                                                  BType expType) {
+
+        boolean areSemanticErrorsPresent = false;
+
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2029]: Shall we extract `((BLangWorkerMultipleReceive.BLangWorkerReceiveField) receiveField)` to a variable and simplify the code. I think we can meet line length requirements also.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -1826,6 +1832,258 @@ public void visit(BLangWorkerReceive workerReceiveExpr) {
...
+            String key;
+            key = receiveField.getWorkerFieldName() != null ?
+                    receiveField.getWorkerFieldName().getValue() : receiveField.getWorkerName().getValue();
+            if (!lhsFields.containsKey(key)) {
+                // Check if the field is sealed, if so you cannot have dynamic fields
+                if (((BRecordType) expType).sealed) {
+                    dlog.error(bLangWorkerMultipleReceive.pos, DiagnosticCode.INVALID_FIELD_NAME_RECORD_LITERAL, key,
+                            expType);
+                    resultType = symTable.semanticError;
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1874]: Can we return here and remove `if (symTable.semanticError != resultType)` in L1889?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -1826,6 +1832,258 @@ public void visit(BLangWorkerReceive workerReceiveExpr) {
...
+        }
+
+        if (receiveField.getWorkerFieldName() != null) {
+            // Type check with worker field name
+            BSymbol fieldSymbol = symResolver.lookupSymbolInMainSpace(
+                    env,
+                    names.fromIdNode((
+                                    (BLangWorkerMultipleReceive.BLangWorkerReceiveField) receiveField).workerFieldName
+                                    ));
+            ((BLangWorkerMultipleReceive.BLangWorkerReceiveField) receiveField).workerFieldName.type
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1937]: shall we extract `((BLangWorkerMultipleReceive.BLangWorkerReceiveField) receiveField)` to a variable here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    public static class BLangWorkerReceiveField implements WorkerReceiveFieldNode {
+
+        public BLangIdentifier workerIdentifier;
+        public BLangIdentifier workerFieldName;
+        public BLangExpression keyExpr;
+        public boolean isChannel = false;
+        public SymbolEnv env;
+        public BType workerType;
+        public BType matchingSendsError; // TODO: Remove this line
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java:121]: Note the TODO

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 5]
```diff
@@ -914,7 +916,13 @@ public void visit(BLangWorkerReceive workerReceiveNode) {
...
     }
 
-    // Expressions
+    @Override
+    public void visit(BLangWorkerMultipleReceive workerMultipleReceive) {
+        for (WorkerMultipleReceiveNode.WorkerReceiveFieldNode receiveField : workerMultipleReceive.receiveFields) {
+            ((BLangWorkerMultipleReceive.BLangWorkerReceiveField) receiveField).sendExpression.accept(this);
+        }
+    }
+// Expressions
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:925]: minor indentation change

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 3 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1937]: Fix the formatting as well. Saw multiple other places as well, with formatting issues

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 6]
```diff
@@ -3173,13 +3338,33 @@ public BLangNode currentAction() {
...
+
         public boolean currentIsReceive(String sourceWorkerId) {
+
             if (this.done()) {
                 return false;
             }
             BLangNode action = this.currentAction();
-            return !isWorkerSend(action) && !isWorkerSyncSend(action) &&
-                    ((BLangWorkerReceive) action).workerIdentifier.value.equals(sourceWorkerId);
+            if (isWorkerMultipleReceive(action)) {
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3361]: Shall we take result of `!isWorkerSend(action) && !isWorkerSyncSend(action)` to a variable before the if condition.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 7]
```diff
@@ -3173,13 +3338,33 @@ public BLangNode currentAction() {
             return this.actions.get(this.currentState);
         }
 
+        private boolean checkAndSetWorkerMultipleReceiveSource(BLangWorkerMultipleReceive multipleReceive,
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3341]: minor suggestion to change `recievedWorker` variable name. `reciev+ed` suggest that receiver has already received. Therefore `receivedWorker.getHasReceived()` feels bit odd.

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-multiple-receive.bal [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        map<int|string> m = <- { c:w1, d:w3 };
+        int x = <- w3;
+        io:println("x is: ", x);
+        io:println("m: ", m);
+        return m;
+    }
+
+    map<int|string> i = wait w2;
+    return i;
+}
```
Comment (rdhananjaya) [inline tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-multiple-receive.bal:183]: Missing newline
And shall we add a test case where the sender may return an error before sending,
following is a similar code written using normal worker receive.
```ballerina
public function main() returns error? {
    worker w0 returns error? {
        if (true) {
            return error("hi");
        }
        3 -> w1;
    }
    worker w1 {
        int|error i = <- w0;
        io:println(i);
    }

    io:println("this is the end");
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed by changing the function argument type of `receiveField` to `BLangWorkerMultipleReceive.BLangWorkerReceiveField`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2105]: If we return here, it wouldn't complete the for-loop and execute the `checkMissingReqFieldsForWait(((BRecordType) expType), rhsFields, waitExpr.pos);`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 3 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed, @pubudu91 fixed the formatting issue in the other places as well

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java (hunk 4 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java]: Removed the TODO.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java (hunk 5 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:927]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 6 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3409]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java [hunk 9]
```diff
@@ -2085,6 +2085,44 @@ public void exitWorkerReceiveExpression(BallerinaParser.WorkerReceiveExpressionC
...
+        this.pkgBuilder.startWorkerMultipleReceive(getCurrentPos(ctx));
+    }
+
+    @Override
+    public void exitWorkerMultipleReceiveExpression(BallerinaParser.WorkerMultipleReceiveExpressionContext ctx) {
+        if (isInErrorState) {
+            return;
+        }
+
+//        this.pkgBuilder.startWorkerMultipleReceive(getCurrentPos(ctx));
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: nice

**Comment by MaryamZi** [comment]
@sanjana, the build seems to be failing. Can you please check and update the PR?

**Comment by sanjana** [comment]
@MaryamZi, thanks for informing. I'll check into it and update the PR.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 7 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3384]: Renamed `recievedWorker` to `receiveField`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        @Override
+        public IdentifierNode getWorkerFieldName() {
+
+            return workerFieldName;
+        }
+
+        @Override
+        public void setHasReceived(boolean value) {
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangWorkerMultipleReceive.java]: Changed to `setCurrentReceiveFieldVisited()`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 11]
```diff
@@ -1936,6 +1941,70 @@ public void visit(BLangWorkerReceive workerReceiveNode) {
...
+
+            if ((sender.tag & SymTag.VARIABLE) != SymTag.VARIABLE) {
+                sender = symTable.notFoundSymbol;
+            }
+
+            verifyPeerCommunication(
+                    workerMultipleReceive.pos,
+                    sender,
+                    workerIdentifier.value
+                                   );
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Done

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 12]
```diff
@@ -1982,10 +1984,37 @@ public BLangNode transform(TrapExpressionNode trapExpressionNode) {
...
+        if (receiveActionNode.receiveWorkers().kind().equals(SyntaxKind.RECEIVE_FIELDS)) {
+            BLangWorkerMultipleReceive workerMultipleReceive = TreeBuilder.createWorkerMultipleReceiveExpressionNode();
+
+            SeparatedNodeList receiveFields = ((ReceiveFieldsNode) receiveActionNode.receiveWorkers()).receiveFields();
+            int size = receiveFields.size();
+            for (int i = 0; i < size; i++) {
+                Node receiveField = receiveFields.get(i);
+                BLangWorkerMultipleReceive.BLangWorkerReceiveField workerReceiveField
+                        = TreeBuilder.createReceiveFieldNode();
+                if (receiveField.kind().equals(SyntaxKind.SIMPLE_NAME_REFERENCE)) {
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java]: Done

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/WorkerMultipleReceiveNode.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+        void setWorkerFieldName(IdentifierNode identifierNode);
+
+        ExpressionNode getKeyExpression();
+
+        IdentifierNode getWorkerName();
+
+        IdentifierNode getWorkerFieldName();
+
+        void setSendExpression(BLangExpression sendExpression);
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/WorkerMultipleReceiveNode.java:52]: Added the newline space.

**Pull request closed without merging**